### PR TITLE
feat: order access-scope regexes by specificity and cache compilations

### PR DESCRIPTION
**Pull request by wernerdweight**
## Summary

- Regex-enabled keys in `AccessScope` / `FUPScope` are now tried in **most-specific-first** order (pattern length descending, lexicographic tie-break), so a narrow deny rule reliably shadows a broader allow rule regardless of map iteration order.
- Compiled regexes are memoized in a process-wide `sync.Map`, so each pattern is compiled at most once per process. Invalid patterns are cached as typed-nil to avoid repeat compile attempts.
- Exact (non-regex) key lookups still win over any regex match.

## Why

Scope lookups previously iterated the underlying `map[string]any` in random order. With overlapping regexes this was non-deterministic — you couldn't, for example, layer

```json
{
  "/token/generate": true,
  "r#^/similarity/v[0-9]+/.*$": true,
  "r#^/similarity/v[0-9]+/(chat|sessions).*$": false
}
```

and trust that `/similarity/v1/chat` would be denied. Now it is: the longer (more specific) regex is tried first, and its `false` value shadows the broader allow.

## What changed

- `auth/contract/entity.go`
  - New helpers: `regexCache` + `getCompiledScopeRegex`, `sortedRegexScopeKeys`, `lookupScopeEntry`.
  - `AccessScope.GetAccessibility`, `FUPScope.GetLimit`, and `FUPScope.HasLimit` rewritten around the new helper. Leaf-type handling (bool/string/int/float32/float64) and nested-scope descent are preserved.
  - Segment walk now uses `for index, segment := range pathSegments`, which also fixes a latent bug in the old code where the inner regex loop's `continue` targeted the wrong loop after descending into a regex-matched nested scope.
- `auth/contract/entity_test.go`
  - 17 new `AccessScope` cases (the `/similarity/...` example, three-regex chains with middle-wins/most-specific-wins/least-specific-wins, exact-match-beats-regex, on-behalf vs deny precedence, invalid regex is skipped, deterministic length ties, nested-scope ordering).
  - 3 new `FUPScope` cases (specific-beats-broad, broad-fallback-when-specific-misses, exact-match precedence).
  - New `TestSortedRegexScopeKeys` — sorter unit tests (length desc + lex tie-break, ignores non-regex keys).
  - New `TestGetCompiledScopeRegex_CachesAndHandlesInvalid` — verifies pointer identity across repeat calls, nil return on invalid patterns, and that failures are cached.

## Notes for reviewers

- **Specificity metric**: pattern length (descending). Simple, deterministic, matches intuition for realistic path patterns. A pathological case like a short-but-heavily-constrained `a{1,100}` won't rank intuitively, but that hasn't come up in practice and we can revisit with a literal-character-count heuristic later.
- **No backtracking**: if a regex matches but its value is a nested scope, the walk commits to that scope — it doesn't back out to try a less-specific regex if later segments fail. Preserves the original behavior.
- **Cache lifetime**: process-wide and unbounded. Pattern strings come from configured scopes, not user input, so cardinality is bounded by config size.
- All existing tests still pass; `go vet ./...` is clean.

## Test plan

- [x] `go test ./auth/contract/...` — all passing (including 20+ new cases)
- [x] `go test ./...` — full suite green
- [x] `go vet ./...` — clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Performance**
  * Access scope and rate limit lookups now utilize pattern caching for faster responses

* **Improvements**
  * Deterministic pattern matching ensures consistent behavior across requests
  * Enhanced error resilience in scope resolution

* **Tests**
  * Comprehensive new test coverage for pattern matching specificity and edge cases

### CODE REVIEW

File: auth/contract/entity_test.go [hunk 1]
```diff
@@ -314,6 +594,37 @@ func TestFUPScope_GetLimit_HasLimit(t *testing.T) {
...
+		{
+			name: "FUP regex ordering: specific regex wins over broad",
+			scope: FUPScope{
+				"r#^/api/v[0-9]+$":          map[string]any{"hourly": 321},
+				"r#^/api/v[0-9]+/premium$":  map[string]any{"hourly": 123},
+			},
+			args: args{path: "/api/v1/premium.hourly"},
+			want: &testValue,
+			has:  true,
+		},
```
[the remark this answers is not in the document]
Comment (wernerdweight) [reply auth/contract/entity_test.go:646]: fixed in [52ca1bd](https://github.com/wernerdweight/api-auth-go/pull/9/commits/52ca1bd3a679217ae4fd3dfdc51907671e020a1d)

File: auth/contract/entity.go [hunk 2]
```diff
@@ -1,12 +1,83 @@
...
+		}
+		return cached.(*regexp.Regexp)
+	}
+	compiled, err := regexp.Compile(pattern)
+	if err != nil {
+		regexCache.Store(pattern, (*regexp.Regexp)(nil))
+		return nil
+	}
+	regexCache.Store(pattern, compiled)
+	return compiled
```
[the remark this answers is not in the document]
Comment (wernerdweight) [reply auth/contract/entity.go]: fixed in [52ca1bd](https://github.com/wernerdweight/api-auth-go/pull/9/commits/52ca1bd3a679217ae4fd3dfdc51907671e020a1d)
